File: examples/python/02_approval_flow.py
"""AGR Approval Flow — full lifecycle with background auto-approve."""
import os
import time
import threading
import httpx
from agr import AGRClient

API_KEY = os.environ.get("AGR_API_KEY", "agr_sk_YOUR_KEY_HERE")
BASE_URL = os.environ.get("AGR_BASE_URL", "http://localhost:8000")

agr = AGRClient(api_key=API_KEY, base_url=BASE_URL)


def auto_approve(approval_id: str, delay: float = 3.0) -> None:
    """Simulate a human approver by auto-approving after a short delay."""
    time.sleep(delay)
    print(f"\n  [auto-approver] Approving {approval_id}...")
    try:
        resp = httpx.post(
            f"{BASE_URL}/v1/approvals/{approval_id}/approve",
            headers={"Authorization": f"Bearer {API_KEY}"},
            json={"comment": "Auto-approved by demo script"},
            timeout=10,
        )
        resp.raise_for_status()
        print(f"  [auto-approver] Done — status {resp.status_code}")
    except Exception as exc:
        print(f"  [auto-approver] Error: {exc}")


def main() -> None:
    print("=== AGR Approval Flow Demo ===\n")

    # Step 1: Evaluate a large fund transfer
    print("--- Step 1: Evaluate transfer_funds ($50,000) ---")
    result = agr.evaluate(
        agent="finance-agent",
        action="transfer_funds",
        resource="bank-account-001",
        context={
            "amount": 50000,
            "destination": "vendor-acct-xyz",
            "destination_country": "IN",
        },
    )
    print(f"  Decision:    {result.decision}")
    print(f"  Approval ID: {result.approval_id}")
    print(f"  Risk Score:  {result.risk_score}")

    if not result.requires_approval:
        print(f"\n  ⚠️  Expected APPROVAL_REQUIRED, got {result.decision}")
        print("  Import finance_controls policies first: bash ../curl/08_import_policies_yaml.sh --commit")
        return

    print("\n✅ APPROVAL_REQUIRED — waiting for human (auto-approve in 3s)...")

    # Step 2: Launch background auto-approver
    print("\n--- Step 2: Starting background approver thread ---")
    approver = threading.Thread(
        target=auto_approve,
        args=(result.approval_id,),
        daemon=True,
    )
    approver.start()

    # Step 3: Wait for approval
    print("\n--- Step 3: Polling for approval decision ---")
    approved = agr.wait_for_approval(result.approval_id, timeout=30)
    approver.join(timeout=5)

    if approved:
        print("\n✅ Transfer approved — agent may now execute")
    else:
        print("\n❌ Transfer rejected or timed out")

    # Step 4: Verify audit trail
    print("\n--- Step 4: Recent audit events ---")
    import httpx as _httpx
    audit_resp = _httpx.get(
        f"{BASE_URL}/v1/audit?limit=5",
        headers={"Authorization": f"Bearer {API_KEY}"},
        timeout=10,
    )
    audit_resp.raise_for_status()
    events = audit_resp.json()
    items = events if isinstance(events, list) else events.get("items", [])
    for event in items[:5]:
        print(f"  [{event.get('event_type','?'):25s}] agent={event.get('agent_id','?')} action={event.get('action','?')}")

    print("\n✅ Approval lifecycle complete")


if __name__ == "__main__":
    main()

File: examples/node/02_approval_flow.ts
/**
 * AGR Approval Flow — full lifecycle in TypeScript.
 * Evaluates a large transfer, auto-approves after 3s, then waits for the decision.
 */
import { AGRClient } from '../../packages/agr-sdk-ts/src/index.js'

const API_KEY = process.env['AGR_API_KEY'] || 'agr_sk_YOUR_KEY_HERE'
const BASE_URL = process.env['AGR_BASE_URL'] || 'http://localhost:8000'

const agr = new AGRClient({ apiKey: API_KEY, baseUrl: BASE_URL })

async function autoApprove(approvalId: string, delayMs = 3000): Promise<void> {
  await new Promise((resolve) => setTimeout(resolve, delayMs))
  console.log(`\n  [auto-approver] Approving ${approvalId}...`)
  const resp = await fetch(`${BASE_URL}/v1/approvals/${approvalId}/approve`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${API_KEY}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ comment: 'Auto-approved by Node.js example' }),
  })
  if (!resp.ok) {
    console.error(`  [auto-approver] HTTP ${resp.status}`)
  } else {
    console.log(`  [auto-approver] Done`)
  }
}

async function main(): Promise<void> {
  console.log('=== AGR Approval Flow Demo (TypeScript) ===\n')

  // Step 1: Evaluate
  console.log('--- Step 1: Evaluate transfer_funds ($50,000) ---')
  const result = await agr.evaluate(
    'finance-agent',
    'transfer_funds',
    'bank-account-001',
    { amount: 50000, destination_country: 'IN' },
  )

  console.log(`  Decision:    ${result.decision}`)
  console.log(`  Approval ID: ${result.approvalId ?? 'N/A'}`)
  console.log(`  Risk Score:  ${result.riskScore}`)

  if (!result.requiresApproval) {
    console.log(`\n  ⚠️  Expected APPROVAL_REQUIRED, got ${result.decision}`)
    console.log('  Import finance_controls policies first:')
    console.log('    bash ../curl/08_import_policies_yaml.sh --commit')
    return
  }

  console.log('\n✅ APPROVAL_REQUIRED — auto-approving in 3 seconds...')

  // Step 2: Start auto-approver in background
  console.log('\n--- Step 2: Starting auto-approver ---')
  const approvalPromise = autoApprove(result.approvalId!)

  // Step 3: Wait for approval
  console.log('\n--- Step 3: Polling for approval decision ---')
  const approved = await agr.waitForApproval(result.approvalId!, {
    pollInterval: 1000,
    timeout: 30000,
  })
  await approvalPromise

  if (approved) {
    console.log('\n✅ Transfer approved — agent may now execute')
  } else {
    console.log('\n❌ Transfer rejected or timed out')
  }

  // Step 4: Show audit events
  console.log('\n--- Step 4: Recent audit events ---')
  const auditResp = await fetch(`${BASE_URL}/v1/audit?limit=5`, {
    headers: { Authorization: `Bearer ${API_KEY}` },
  })
  if (auditResp.ok) {
    const audit = await auditResp.json() as { items?: Array<Record<string, unknown>> } | Array<Record<string, unknown>>
    const items = Array.isArray(audit) ? audit : (audit.items ?? [])
    for (const event of items.slice(0, 5)) {
      const type = String(event['event_type'] ?? '?').padEnd(25)
      console.log(`  [${type}] agent=${event['agent_id']} action=${event['action']}`)
    }
  }

  console.log('\n✅ Approval lifecycle complete')
}

main().catch(console.error)
